node/src/components/consensus/highway_core/state.rs

```diff
@@ -200,9 +200,11 @@ pub(crate) struct State<C: Context> {
     /// Panoramas that the protocol observed.
     panoramas: Panoramas<C>,
     /// All currently endorsed units, by hash.
-    endorsements: HashMap<C::Hash, Vec<SignedEndorsement<C>>>,
+    endorsements: HashMap<C::Hash, ValidatorMap<Option<C::Signature>>>,
     /// Units that don't yet have 2/3 of stake endorsing them.
-    incomplete_endorsements: HashMap<C::Hash, Vec<SignedEndorsement<C>>>,
+    /// Signatures are stored in a map so that a single validator sending lots of signatures for
+    /// different units doesn't cause us to allocate a lot of memory.
+    incomplete_endorsements: HashMap<C::Hash, BTreeMap<ValidatorIndex, C::Signature>>,
```

Comment:
I am not sure I understand – `incomplete_endorsements` is keyed by a unit's hash, so there can be only `validators_count` entries per each unit. A single validator sending lots of signatures for different units would create more entries in the outer collection, not inner. Why not use `HashMap<Hash, ValidatorMap<Signature>>`?

Reply:
Because a `ValidatorMap` is a `Vec` internally, whose length is the number N of validators, so it always uses as much space as N endorsements even if just one of them is present.

Reply:
Right, that's the reason I didn't use it originally as well. I was on the fence, because eventually it will have to be of at lest `0.5 * validator_count` size. And how does the `BTreeMap` work internally? It it of size `1` at the beginning, and after adding one element it will grow to size 2, then 4, 8, 16 etc? Is this "resizing" negligible? I think I prefer the consistency (`ValidatorMap`) and I am not convinced the performance gains are real.

Also, is the comment necessary?

Reply:
I think `BTreeMap`s don't even double their size, unlike hash maps: Every operation is always `O(log(n))`.

It's probably not a real "attack", but if we use a `ValidatorMap` or a `Vec` with sufficient capacity, a single validator can, by endorsing everything, waste N times as much of our memory as with the `BTreeMap`. And since almost none of our computations here are constant time anyway, I don't see the benefit of the consistency in this case.

Happy to remove the comment, if you prefer. I just thought the reader would wonder why we use `ValidatorMap` in `endorsements` but not in `incomplete_endorsements`.

Reply:
OK, maybe you're right. Let's leave it.